test/helpers/position_assertions.cc

```diff
@@ -1218,6 +1223,162 @@ string ApplyCompletionAssertion::toString() const {
+}
+
+ApplyRenameAssertion::ApplyRenameAssertion(string_view filename, unique_ptr<Range> &range, int assertionLine,
+                                           string_view version, string newName, bool invalid,
+                                           string expectedErrorMessage)
+    : RangeAssertion(filename, range, assertionLine), version(string(version)), newName(newName), invalid(invalid),
+      expectedErrorMessage(expectedErrorMessage) {}
+
+void ApplyRenameAssertion::checkAll(const vector<shared_ptr<RangeAssertion>> &assertions,
+                                    const UnorderedMap<string, shared_ptr<core::File>> &sourceFileContents,
+                                    LSPWrapper &wrapper, int &nextId, string errorPrefix) {
+    for (auto assertion : assertions) {
+        if (auto assertionOfType = dynamic_pointer_cast<ApplyRenameAssertion>(assertion)) {
+            assertionOfType->check(sourceFileContents, wrapper, nextId, errorPrefix);
+        }
+    }
+}
+
+void ApplyRenameAssertion::check(const UnorderedMap<std::string, std::shared_ptr<core::File>> &sourceFileContents,
+                                 LSPWrapper &wrapper, int &nextId, std::string errorPrefix) {
+    auto prepareRenameResponse = doTextDocumentPrepareRename(wrapper, *this->range, nextId, this->filename);
+    // TODO: test placeholder in prepareRenameResponse
+
+    // A rename at an invalid position
+    if (newName.empty() && invalid) {
+        REQUIRE_EQ(prepareRenameResponse, nullptr);
+    } else {
+        REQUIRE_NE(prepareRenameResponse, nullptr);
+    }
+
+    auto workspaceEdits =
+        doTextDocumentRename(wrapper, *this->range, nextId, this->filename, newName, expectedErrorMessage);
+    // A rename at a valid position but with an invalid new name
+    if (invalid) {
+        REQUIRE_EQ(workspaceEdits, nullptr);
+        return;
+    }
+
+    {
+        INFO("doTextDocumentRename failed; see error above.");
+        REQUIRE_NE(workspaceEdits, nullptr);
+    }
+
+    auto &renameItems = *workspaceEdits->documentChanges;
+    auto it = sourceFileContents.find(this->filename);
+    {
+        INFO(fmt::format("Unable to find source file `{}`", this->filename));
+        REQUIRE_NE(it, sourceFileContents.end());
+    }
+
+    size_t index = this->filename.rfind("/", this->filename.length());
+    auto testDataPath = this->filename.substr(0, index);
+    // map of all .rb and .rbi files
+    UnorderedMap<string, string> sourceFiles;
+    // map of all .rbedited files whose version equals `this->version`
+    UnorderedMap<string, string> expectedEditedFiles;
+    for (auto filePath : FileOps::listFilesInDir(testDataPath, {".rb", ".rbi", ".rbedited"}, false, {}, {})) {
+        auto extension = FileOps::getExtension(filePath);
+        if (extension == "rbedited") {
+            auto extensionIndex = filePath.rfind(".rbedited", filePath.length());
+            auto fileVersion = string(FileOps::getExtension(filePath.substr(0, extensionIndex)));
```

Comment:
Ah, so this caused an issue before because you are:

1. Calling `filePath.substr`, which creates a new `string` (not a `string_view` of `filePath`
2. `FileOps::getExtension` is returning a `string_view` of that new string
3. The new string is destructed after this line since it's not stored anywhere.

The new call to `string` seems to properly copy the string before the destruction.